pkg/ip/link.go

```diff
@@ -172,28 +172,41 @@ func DelLinkByNameAddr(ifName string, family int) (*net.IPNet, error) {
 	return addrs[0].IPNet, nil
 }
 
+// SetHWAddrByIP sets the hardware address of an interface/link. If an IPv4
+// address is provided (for IPv4-only or dual-stack operation), then the
+// that IPv4 address is assumed to be locally unique, and the hardware
+// address that is configured comprises the 4 bytes of IPv4 address preceded
+// by a hard-coded 2-byte prefix (0a:58). If only an IPv6 address is
+// provided, then the last 4 bytes of the interface/link's current hardware
+// address are assumed to be random, and the hardware address that is
+// configured is derived by overwriting the first two bytes of the
+// current hardware address with a different hard-coded prefix (6a:58).
 func SetHWAddrByIP(ifName string, ip4 net.IP, ip6 net.IP) error {
 	iface, err := netlink.LinkByName(ifName)
 	if err != nil {
 		return fmt.Errorf("failed to lookup %q: %v", ifName, err)
 	}
 
+	var hwAddr net.HardwareAddr
 	switch {
-	case ip4 == nil && ip6 == nil:
-		return fmt.Errorf("neither ip4 or ip6 specified")
-
 	case ip4 != nil:
-		{
-			hwAddr, err := hwaddr.GenerateHardwareAddr4(ip4, hwaddr.PrivateMACPrefix)
-			if err != nil {
-				return fmt.Errorf("failed to generate hardware addr: %v", err)
-			}
-			if err = netlink.LinkSetHardwareAddr(iface, hwAddr); err != nil {
-				return fmt.Errorf("failed to add hardware addr to %q: %v", ifName, err)
-			}
+		hwAddr, err = hwaddr.GenerateHardwareAddr4(ip4, hwaddr.PrivateMACPrefix)
+		if err != nil {
+			return fmt.Errorf("failed to generate hardware addr: %v", err)
 		}
 	case ip6 != nil:
-		// TODO: IPv6
+		attrs := iface.Attrs()
+		hwAddr, err = hwaddr.GenerateHardwareAddr6(attrs.HardwareAddr,
+			hwaddr.PrivateMACPrefix6)
+		if err != nil {
+			return fmt.Errorf("failed to generate hardware addr: %v", err)
+		}
+	default:
+		return fmt.Errorf("neither ip4 or ip6 specified")
```

Comment:
"unable to generate hardware address as neither IPv4 nor IPv6 address specified"?